design/hostclaim-multitenancy.md

```diff
@@ -0,0 +1,582 @@
+
+We need to relax the constraint that the cluster and the BareMetalHosts are
+in the same namespace but we also need a solution that gives sufficient control
+and visibility over the workload deployed on those servers so that tenants
+can maintain the level of information they have had so far.
+
+This proposal tackles the objective through three changes:
+
+* it introduces a new resource called HostClaim that decouples
+  the definition of the workload performed by
+  the Metal3 machine controller from the actual compute resource.
+  This resource acts as a security boundary.
+* it uses the HardwareData resource as the only visible resource for end-users
+  containing both hardware details and metadata used for selection and templating.
+* it introduces a HostDeployPolicy resource that describes the namespaces that contain
+  the HostClaims authorized to be bound to BareMetalHosts in the namespace of the
+  HostDeployPolicy resource.
+
+### Goals
+
+* Split responsibilities between infrastructure teams (role *infrastructure manager*),
+  who manage servers, and cluster administrators (role *cluster manager*),
+  who create/update/scale bare-metal clusters deployed
+  on those servers, using traditional Kubernetes RBAC to ensure isolation.
+* Define a resource where a user can request a compute resource to execute
+  an arbitrary workload described by an OS image and an initial configuration.
+  The user (role *end-user*, a cluster manager is an end-user) does not need
+  to know exactly which BareMetalHost is used and does not control its BMC.
+* Using BareMetalHosts defined in other clusters. Support as described here
+  will be limited to the use with cluster-api as it will be handled at the
+  level of the Metal3Machine.
+
+### Non-Goals
+
+* Compute resource quotas. The HostClaim resource should make it possible to
+  develop a framework to limit the number/size of compute resources allocated
+  to a tenant, similar to how quotas work for pods. However, the specification
+  of such a framework will be addressed in another design document.
+* Implementing network isolation between tenant clusters using a common set
+  of BareMetalHosts.
+
+## Proposal
+
+### User Stories
+
+#### Deployment of Simple Workloads
+
+##### User Point of View
+
+As an *end-user*, I would like to execute a workload on an arbitrary server.
+
+The OS image is available in qcow format on a remote server at ``url_image``.
+It supports cloud-init and a script can launch the workload at boot time
+(e.g., a systemd service).
+
+The cluster offers bare-metal as a service using Metal3 baremetal-operator.
+However, as a regular user, I am not allowed to directly access the definitions
+of the servers, but I can read the result of the inspection (HardwareData).
+The HardwareData resource also contains a copy of the metadata of the
+BareMetalHost visible to the end-user.
+
+All HardwareData are labeled with an ``infra-kind`` label whose
```

Comment:
This is a manual action by the admin, right? Worth clarifying the actors IMO.

Reply:
Yes and as a consequence of previous comment it should be put on the baremetalhost. HardwareData is still seen as a transient resource. I will add a sentence defining the hardware administrator.